lib/ts/Diags.cc

```diff
@@ -816,29 +820,39 @@ Diags::set_stdout_output(const char *_bind_stdout)
   if (strcmp(_bind_stdout, "") == 0)
     return false;
 
-  if (stdout_log) {
-    delete stdout_log;
-    stdout_log = NULL;
-  }
-
-  // create backing BaseLogFile for stdout
-  stdout_log = new BaseLogFile(_bind_stdout);
+  BaseLogFile *old_stdout_log = stdout_log;
+  BaseLogFile *new_stdout_log = new BaseLogFile(_bind_stdout);
 
   // on any errors we quit
-  if (!stdout_log || stdout_log->open_file() != BaseLogFile::LOG_FILE_NO_ERROR) {
+  if (!new_stdout_log || new_stdout_log->open_file() != BaseLogFile::LOG_FILE_NO_ERROR) {
     fprintf(stderr, "[Warning]: unable to open file=%s to bind stdout to\n", _bind_stdout);
-    delete stdout_log;
+    fprintf(stderr, "[Warning]: stdout is currently not bound to anything\n");
+    delete new_stdout_log;
+    lock();
     stdout_log = NULL;
+    unlock();
     return false;
   }
-  if (!stdout_log->m_fp) {
+  if (!new_stdout_log->m_fp) {
     fprintf(stderr, "[Warning]: file pointer for stdout %s = NULL\n", _bind_stdout);
-    delete stdout_log;
+    fprintf(stderr, "[Warning]: stdout is currently not bound to anything\n");
+    delete new_stdout_log;
+    lock();
     stdout_log = NULL;
+    unlock();
     return false;
   }
 
-  return rebind_stdout(fileno(stdout_log->m_fp));
+  // now exchange the stdout_log pointer
+  lock();
+  stdout_log = new_stdout_log;
+  bool ret = rebind_stdout(fileno(new_stdout_log->m_fp));
```

Comment:
I think there's a couple of issues here. First, if the error can't (or shouldn't) reasonably be handled by the caller, then there's little value in trying to propagate it. 

``fileno()`` can't fail unless there's a programmer error. and `dup2()`` can't fail unless there is a programmer error or we run out of file descriptors. If you look at ``rebind_stderr()``, we don't even propagate the return value of ``dup2()``.

I think that this should be handled by a ``ink_release_assert``.